### PR TITLE
Add finalizers to ensure that repos are closed and blobreaders are closed (#19495)

### PR DESCRIPTION
**Pull request by zeripath**
Backport #19495

It may be prudent to add runtime finalizers to the git.Repository and
git.blobReader objects to absolutely ensure that these are both properly
cancelled, cleaned and closed out.

This commit is a backport of an extract from #19448

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Review by 6543 — APPROVED** [review]

File: modules/git/repo_base_gogit.go [hunk 1]
```diff
@@ -63,13 +63,17 @@ func OpenRepositoryCtx(ctx context.Context, repoPath string) (*Repository, error
...
+	repo := &Repository{
 		Path:         repoPath,
 		gogitRepo:    gogitRepo,
 		gogitStorage: storage,
 		tagCache:     newObjectCache(),
 		Ctx:          ctx,
-	}, nil
+	}
+
+	runtime.SetFinalizer(repo, func(repo *Repository) { repo.Close() })
```
Comment (Gusted) [inline modules/git/repo_base_gogit.go]: You forgot to import the `runtime` library

**Review by Gusted — COMMENTED** [review]
In the `Close` function we want to set the the Finalizer of that object to `nil`.

**Comment by lunny** [comment]
I don't think it's a good idea.

**Review by a1012112796 — APPROVED** [review]

**Comment by wxiaoguang** [comment]
For 1.16, it might work and might resolve some resource leaking problems.

For 1.17, I am a little conservative about it. The reason is that GC time can not be controlled (well, should not be controlled), and sometimes the leaked resource may be still referenced by other variables while the resource itself should be closed. So I think it's safer and proper to fix these leaked resources if there is a chance. The runtime finalizers might help developers to find these leaked resources.

**Review by wxiaoguang — APPROVED** [review]

**Comment by zeripath** [comment]
Please note, I have no information to suggest that we have any leaks, nor that we should consider that relying on finalizers is a good idea.

However, if we do have leaks - it is infinitely better for a finalizer to take care of them than to have a leak there.

**Comment by wxiaoguang** [comment]
> Please note, I have no information to suggest that we have any leaks, nor that we should consider that relying on finalizers is a good idea.

If no leak, why the finalizers are needed? Close a already closed object?

> However, if we do have leaks - it is infinitely better for a finalizer to take care of them than to have a leak there.

I also agree that finalizers can help to prevent leaking. I mean if there are some leaks, we need logs to tell developers that there is something incorrect.

**Comment by 6543** [comment]
hop or top - should we move or merge?!

**Comment by wxiaoguang** [comment]
Vote to merge

**Comment by Gusted** [comment]
Agree to merge this. It doesn't any harm to merge and is a nice safety measurement.

**Comment by ashimokawa** [comment]
This PR introduced a memory leak on codeberg.org leading to about 25GB of leaked memory in 2 days. We had to revert it.

I think the real cause should be fixed before the next release.

**Pull request merged**